### ISSUE TITLE
Problem installing from GitHub owing to old setuptools on CentOS 

### ISSUE THREAD

**Issue by verdurin** [gc3-uzh-ch/elasticluster#227]
I'm trying to install from GitHub, as usual. I did notice a recent commit stating that the Ansible requirement would remain at 1.9.x for now, but `pip install ansible` of course installed 2.0.0.2.

Inside the virtualenv, I ran `python setup.py install` and saw the following error:

```
(elasticluster)[centos@vanloo-admin src]$ python setup.py install
error in elasticluster setup command: 'install_requires' must be a string or list of strings containing valid project/version requirement specifiers
```

**Comment by riccardomurri** [comment]
What branch are you installing from?

**Comment by verdurin** [comment]
This is with master, which I believe is supposed to be the more stable branch with the way you're managing things now?

**Comment by riccardomurri** [comment]
I cannot reproduce the issue; `master` installs fine in a freshly
created virtualenv on my machine::

    $ git checkout master
    Already on 'master'
    Your branch is up-to-date with 'origin/master'.

    $ git rev-parse --short HEAD
    89958da

    $ mkvirtualenv elasticluster-issue-227
    Using real prefix '/usr'
    New python executable in elasticluster-issue-227/bin/python2
    Also creating executable in elasticluster-issue-227/bin/python
    Installing setuptools, pip, wheel...done.

    (elasticluster-issue-227)$ python setup.py install
    /home/rmurri/.virtualenvs/elasticluster-issue-227/local/lib/python2.7/site-packages/setuptools/dist.py:285: UserWarning: Normalizing '1.2.1.rc1' to '1.2.1rc1'
      normalized_version,
    running install
    running bdist_egg
    running egg_info
    writing requirements to elasticluster.egg-info/requires.txt
    writing elasticluster.egg-info/PKG-INFO
    [...]
    Using /home/rmurri/.virtualenvs/elasticluster-issue-227/lib/python2.7/site-packages
    Finished processing dependencies for elasticluster==1.2.1rc1

    (elasticluster-issue-227)$ which elasticluster
    /home/rmurri/.virtualenvs/elasticluster-issue-227/bin/elasticluster

    (elasticluster-issue-227)$ elasticluster --version
    elasticluster version 1.2.1rc1

No problem is detected by the CI server, either: https://travis-ci.org/gc3-uzh-ch/elasticluster/jobs/103839621

Is it possible that your Git checkout is not up-to-date or that you
have local modifications?

**Comment by verdurin** [comment]
Well, I would say that the procedure you've followed there is rather different than what's in the documentation.

**Comment by riccardomurri** [comment]
> Well, I would say that the procedure you've followed there is rather different than what's in the documentation.

How is it different? The docs tell you to:

1. Create a virtualenv and activate it
2. Check out the git repo
3. Run `python setup.py install`

The only differences I can see is that I used `mkvirtualenv` instead of plain `virtualenv` + `source .../activate` and that I have the git repo already checked out so I just switched branch...

**Comment by verdurin** [comment]
Just tried it again, on a CentOS 7.2 VM:

```
[centos@vanloo-admin src]$ mkdir venv-test
[centos@vanloo-admin src]$ cd venv-test
[centos@vanloo-admin venv-test]$ virtualenv elasticluster
New python executable in elasticluster/bin/python
Installing Setuptools..............................................................................................................................................................................................................................done.
Installing Pip.....................................................................................................................................................................................................................................................................................................................................done.
[centos@vanloo-admin venv-test]$ . elasticluster/bin/activate
(elasticluster)[centos@vanloo-admin venv-test]$ cd elasticluster
(elasticluster)[centos@vanloo-admin elasticluster]$ git clone git://github.com/gc3-uzh-ch/elasticluster.git src
Cloning into 'src'...
remote: Counting objects: 5591, done.
remote: Compressing objects: 100% (29/29), done.
remote: Total 5591 (delta 9), reused 0 (delta 0), pack-reused 5561
Receiving objects: 100% (5591/5591), 2.32 MiB | 229.00 KiB/s, done.
Resolving deltas: 100% (3250/3250), done.
(elasticluster)[centos@vanloo-admin elasticluster]$ cd src
(elasticluster)[centos@vanloo-admin src]$ python setup.py install
error in elasticluster setup command: 'install_requires' must be a string or list of strings containing valid project/version requirement specifiers
```

**Comment by riccardomurri** [comment]
Anyway, just to be sure, here are the steps *exactly as written in the
docs*.  Same outcome, i.e., no error.

    $ virtualenv elasticluster
    New python executable in elasticluster/bin/python
    Installing setuptools, pip, wheel...done.

    $ . elasticluster/bin/activate

    (elasticluster)$ pip install ansible
    Collecting ansible
      Downloading ansible-2.0.0.2.tar.gz (1.5MB)
      [...]
    Successfully installed MarkupSafe-0.23 PyYAML-3.11 ansible-2.0.0.2 ecdsa-0.13 jinja2-2.8 paramiko-1.16.0 pycrypto-2.6.1

    (elasticluster)$ cd elasticluster

    (elasticluster)$ git clone git://github.com/gc3-uzh-ch/elasticluster.git src
    Cloning into 'src'...
    remote: Counting objects: 5591, done.
    remote: Compressing objects: 100% (29/29), done.
    remote: Total 5591 (delta 9), reused 0 (delta 0), pack-reused 5561
    Receiving objects: 100% (5591/5591), 2.32 MiB | 3.01 MiB/s, done.
    Resolving deltas: 100% (3250/3250), done.
    Checking connectivity... done.

    (elasticluster)$ python setup.py install
    /tmp/elasticluster/local/lib/python2.7/site-packages/setuptools/dist.py:285: UserWarning: Normalizing '1.2.1.rc1' to '1.2.1rc1'
      normalized_version,
    running install
    running bdist_egg
    running egg_info
    creating elasticluster.egg-info
    writing requirements to elasticluster.egg-info/requires.txt
    writing elasticluster.egg-info/PKG-INFO
    [....]
    Using /tmp/elasticluster/lib/python2.7/site-packages
    Finished processing dependencies for elasticluster==1.2.1rc1

    (elasticluster)$ which elasticluster
    /tmp/elasticluster/bin/elasticluster

    (elasticluster)$ elasticluster --version
    elasticluster version 1.2.1rc1

**Comment by verdurin** [comment]
Could be a version problem? Here's what I have:

`python-virtualenv-1.10.1-2.el7.noarch`

**Comment by verdurin** [comment]
Okay, it's the specification of the ansible that's the problem.

If I change it to:

```
'ansible>=1.9.4, < 2.0.0.2',
```

for example, it works.

**Comment by riccardomurri** [comment]
You mean `ansible ~= 1.9.4` is the problem?

What version of Python's `setuptools` does CentOS 7 have installed?

**Comment by verdurin** [comment]
Yes.

`python-setuptools-0.9.8-4.el7.noarch`

**Comment by verdurin** [comment]
I've submitted a PR that should fix this. You may wish to fine-tune the version specification.

**Comment by riccardomurri** [comment]
Does it work if you replace the version specification with the following?

    ansible == 1.9.*

**Comment by verdurin** [comment]
No - the error reappears.

**Comment by riccardomurri** [comment]
So the issue is that `setuptools` is too old... the reference implementation for PEP 440 version specifiers was [merged on August 27, 2014][1], whereas [setuptools-0.9.8 was released in summer 2013][2].

[1]: https://github.com/pypa/packaging/pull/1
[2]: https://pypi.python.org/pypi/setuptools/0.9.8

**Comment by riccardomurri** [comment]
So I guess I'll do the following:

- merge your PR for the 1.2.1 release
- try to bootstrap setuptools to a newer release (we need >=8.0 for version specifiers to work) for other branches; running with such an old setuptools might land us other unexpected problems as well...

Does this sound reasonable?

**Comment by verdurin** [comment]
Yes, that makes sense.

**Comment by verdurin** [comment]
Using 

```
pip install --upgrade pip
```

inside the virtualenv works, so that could be added to the instructions, specifically for CentOS.

**Comment by riccardomurri** [comment]
Can you please try installing branch hotfix/ensure-recent-setuptools with the vanilla instructions? (i.e., no `pip install --upgrade` trick)

Does it work?

**Comment by verdurin** [comment]
Yes, that works.

It installs correctly and I can list my existing cluster.

**Comment by riccardomurri** [comment]
Thanks, then I'll use that variant for the development branch

**Comment by riccardomurri** [comment]
Fixed in commit 295c617 on the "master" branch, and in commits b2f782c..7d7e5e5 on the "develop" branch.
